File: Leetcode-solution-c-c++/reducingDishes.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int maxSatisfaction(vector<int> &satisfaction)
    {
        int maxSatisfaction = 0, currSatisfaction = 0, totalSatisfaction = 0;
        sort(satisfaction.begin(), satisfaction.end(), greater<int>());
        for (int num : satisfaction)
        {
            currSatisfaction += num;
            totalSatisfaction += currSatisfaction;
            maxSatisfaction = max(maxSatisfaction, totalSatisfaction);
        }
        return maxSatisfaction;
    }
};

File: Leetcode-solution-python/reducingDishes.py
from typing import List


class Solution:
    def maxSatisfaction(self, satisfaction: List[int]) -> int:
        max_satisfaction = curr_satisfaction = total_satisfaction = 0
        satisfaction.sort(reverse=True)

        for num in satisfaction:
            curr_satisfaction += num
            total_satisfaction += curr_satisfaction
            if total_satisfaction > max_satisfaction:
                max_satisfaction = total_satisfaction

        return max_satisfaction
